crates/store/src/db/sql/utils.rs

```diff
@@ -0,0 +1,154 @@
+use miden_objects::{
+    accounts::{Account, AccountDelta},
+    crypto::hash::rpo::RpoDigest,
+    notes::Nullifier,
+    utils::Deserializable,
+};
+use rusqlite::{
+    params, params_from_iter,
+    types::{Value, ValueRef},
+    Connection, OptionalExtension, ToSql, Transaction,
+};
+
+use crate::errors::DatabaseError;
+
+/// Returns the high 16 bits of the provided nullifier.
+pub fn get_nullifier_prefix(nullifier: &Nullifier) -> u32 {
+    (nullifier.most_significant_felt().as_int() >> 48) as u32
+}
+
+/// Checks if a table exists in the database.
+pub fn table_exists(conn: &Connection, table_name: &str) -> rusqlite::Result<bool> {
+    Ok(conn
+        .query_row(
+            "SELECT 1 FROM sqlite_master WHERE type = 'table' AND name = $1",
+            params![table_name],
+            |_| Ok(()),
+        )
+        .optional()?
+        .is_some())
+}
+
+/// Returns the schema version of the database.
+pub fn schema_version(conn: &Connection) -> rusqlite::Result<usize> {
+    conn.query_row("SELECT * FROM pragma_schema_version", [], |row| row.get(0))
+}
+
+/// Generates a simple insert SQL statement with values for the provided table, fields, and record
+/// number.
+pub fn insert_sql(table: &str, fields: &[&str], record_count: usize) -> String {
+    assert!(record_count > 0);
+
+    format!(
+        "INSERT INTO {table} ({}) VALUES {}",
+        fields.join(", "),
+        format!("({}), ", "?, ".repeat(fields.len()).trim_end_matches(", "))
+            .repeat(record_count)
+            .trim_end_matches(", ")
+    )
+}
+
+/// Generates and executes a bulk insert SQL statement for the provided table, fields, and values.
+///
+/// # Notes
+///
+/// Values are expected to be in the same order as the fields.
+pub fn bulk_insert(
+    transaction: &Transaction,
+    table: &str,
+    fields: &[&str],
+    record_count: usize,
+    values: impl IntoIterator<Item: ToSql>,
+) -> rusqlite::Result<usize> {
+    if record_count == 0 {
+        return Ok(0);
+    }
+
+    let sql = insert_sql(table, fields, record_count);
+
+    transaction.execute(&sql, params_from_iter(values))
+}
+
+/// Converts a `u64` into a [Value].
+///
+/// Sqlite uses `i64` as its internal representation format. Note that the `as` operator performs a
+/// lossless conversion from `u64` to `i64`.
+pub fn u64_to_value(v: u64) -> Value {
+    Value::Integer(v as i64)
+}
+
+/// Converts a `u32` into a [Value].
+///
+/// Sqlite uses `i64` as its internal representation format.
+pub fn u32_to_value(v: u32) -> Value {
+    let v: i64 = v.into();
+    Value::Integer(v)
+}
+
+/// Gets a `u64` value from the database.
+///
+/// Sqlite uses `i64` as its internal representation format, and so when retrieving
+/// we need to make sure we cast as `u64` to get the original value
+pub fn column_value_as_u64<I: rusqlite::RowIndex>(
+    row: &rusqlite::Row<'_>,
+    index: I,
+) -> rusqlite::Result<u64> {
+    let value: i64 = row.get(index)?;
+    Ok(value as u64)
+}
```

Comment:
Are these really necessary? rusqlite already [provides](https://docs.rs/rusqlite/latest/rusqlite/types/trait.ToSql.html) these out the box I think. Not really part of this PR though, so feel free to ignore.

Reply:
I think, we can get rid of `u32_to_value`, but we still need the rest, because `rusqlite` can convert values only if they fit to destination type. It will return errors, if source type under-/overflows destination one.